### ISSUE TITLE
Ability to skip invalidating directories for `touched` children

### ISSUE THREAD

**Issue by stuhood** [facebook/watchman#305]
With watchman `4.5.0` on OSX and a query like:
```
expression=[
  'allof',  # All of the below rules must be true to match.
  ['not', ['pcre', r'^\..*', 'wholename']]  # Exclude files in hidden dirs.
]
```
... `touch`ing an _existing_ file `b/1` contained in a directory `b/` will invalidate the directory in addition to the file. Presumably this is because a directory listing that includes access/modification times might have changed?

It would be great to be able to (even on a best-effort basis) include something in triggered events to indicate _how_ a directory has changed. Our system does not use access/modification times, and thus doesn't need to invalidate directory listings for them.

Thanks for building a great system!

**Comment by wez** [comment]
I wrote up a pretty long explanation of why we don't offer this here: http://stackoverflow.com/a/35753594/149111

With that out of the way, let's focus on helping you with your use case.  Can you tell me a bit more about it?  What I thought you were trying to do based on the description in this issue doesn't quite match up to the way I read the description in the pants issue you linked; I'm almost certainly missing something important :-)

Can you give some more detail on the full query that you're passing to watchman?  I'm particularly interested in how you work with the *since* generator and the *clock* values that watchman passes back.  In theory we can add a *since expression term* (**not** the same thing as the *since generator*!) to also constrain the results by *ctime*.  I think the trick will be in phrasing that filter without you taking watchman implementation details into your application.

**Comment by stuhood** [comment]
Thanks for the quick response!

> doesn't quite match up to the way I read the description in the pants issue you linked

Yea, the linked issue is where I discovered this issue, rather than an expansion of this issue.

> Can you give some more detail on the full query that you're passing to watchman? 
The fully query (before I began to attempt to remove the filtering of directories) [is here](https://github.com/pantsbuild/pants/blob/master/src/python/pants/pantsd/service/fs_event_service.py#L62-L85), but I think that the minimization I included here is pretty straightforward.

The usecase is: we never use file timestamps. And so we'd like to avoid invalidating the parent directory when you touch an _existing_ file (the "existing" bit was missing from the original description: have added it now).

**Comment by wez** [comment]
Can you phrase that in terms of what information you want from watchman and in which circumstances?  invalidation sounds like it is specific to some pants-internal data structure that I'm not familiar with.  The linked issue sounds like you may be computing a directory listing using a glob and that you recompute based on a change notification from watchman; is my interpretation accurate?

At a higher level, is the goal "compute a directory listing using a set of globs as the tree changes"?
Do you use the file change notification stream for more than just this?

Is it fair to say that you'd like to receive all file change notifications, but only directory notifications for deleted dirs?  Are you filtering out the dirs today because you need to perform a relatively expensive glob recomputation?

I know I'm asking a lot of questions: it will help me suggest how best to make use of watchman, and how we might need to make changes to better meet your needs.

**Comment by stuhood** [comment]
Will answer your questions in a second, but first I want to try to demonstrate the problem, as I think it is potentially simple.

This session indicates that when I touch an existing file in a directory on OSX, the "access" time changes, but the modify/change times do not. So it's possible that watchman is triggering an event for access time changes on directories?

```
[stuhood@tw ~/Desktop/stats]$ stat -x 3rdparty
  File: "3rdparty"
  Size: 68           FileType: Directory
  Mode: (0770/drwxrwx---)         Uid: (  501/ stuhood)  Gid: (   20/   staff)
Device: 1,4   Inode: 248704863    Links: 2
Access: Tue Jul 19 12:04:37 2016
Modify: Tue Jul 19 12:04:37 2016
Change: Tue Jul 19 12:04:37 2016
[stuhood@tw ~/Desktop/stats]$ cd 3rdparty/
[stuhood@tw ~/Desktop/stats/3rdparty]$ touch BUILD
[stuhood@tw ~/Desktop/stats/3rdparty]$ stat -x BUILD
  File: "BUILD"
  Size: 0            FileType: Regular File
  Mode: (0660/-rw-rw----)         Uid: (  501/ stuhood)  Gid: (   20/   staff)
Device: 1,4   Inode: 248704897    Links: 1
Access: Tue Jul 19 12:04:52 2016
Modify: Tue Jul 19 12:04:52 2016
Change: Tue Jul 19 12:04:52 2016
[stuhood@tw ~/Desktop/stats/3rdparty]$ stat -x .
  File: "."
  Size: 102          FileType: Directory
  Mode: (0770/drwxrwx---)         Uid: (  501/ stuhood)  Gid: (   20/   staff)
Device: 1,4   Inode: 248704863    Links: 3
Access: Tue Jul 19 12:04:37 2016
Modify: Tue Jul 19 12:04:52 2016
Change: Tue Jul 19 12:04:52 2016
[stuhood@tw ~/Desktop/stats/3rdparty]$ touch BUILD
[stuhood@tw ~/Desktop/stats/3rdparty]$ stat -x .
  File: "."
  Size: 102          FileType: Directory
  Mode: (0770/drwxrwx---)         Uid: (  501/ stuhood)  Gid: (   20/   staff)
Device: 1,4   Inode: 248704863    Links: 3
Access: Tue Jul 19 12:05:19 2016
Modify: Tue Jul 19 12:04:52 2016
Change: Tue Jul 19 12:04:52 2016
```

----

> At a higher level, is the goal "compute a directory listing using a set of globs as the tree changes"?

Correct, but that is a few levels of abstraction up. The fundamental operations are described below.

> Do you use the file change notification stream for more than just this?

We additionally invalidate file fingerprints/content that we hold in memory.

> Is it fair to say that you'd like to receive all file change notifications, but only directory notifications for deleted dirs?

No: we would want to receive an event for either creation, deletion, or child-changes of directories. We don't need to be able to distinguish between these events.

> Are you filtering out the dirs today because you need to perform a relatively expensive glob recomputation?

The original goal of filtering out the directories was to avoid the bug described by this (#305) issue. It seems very likely that there is a simple solution to this issue (#305) if we're able to agree on its definition.

----

In summary: pants implements all other operations in terms of the following operations:

1. Get a DirectoryListing for a Dir - children and their types (but _not_ their "clocks")
2. Get FileContent for a File - content of a file (but _not_ its "clock")
3. Get a ReadLink for a Link - read the path represented by a link, non-recursively

We use watchman to watch for events that affect the results of those individual operations (and then we transitively invalidate things that depended on the operations).

The issue described here is that touching an existing regular file triggers an event for its parent directory as well. It's not clear why this fires, as changing the mod/access time of a child file should not affect the mod/access time of the parent directory.

**Comment by wez** [comment]
Thanks for that explanation.  It does sound like a bug and I think I can see the cause; I've attached a pull request that I think will take care of this.  Are you in a position to test it out locally to confirm that it helps you?

Assuming that you have a working build environment, the easiest way to do this is something like this:

```
<clone watchman>
./autogen.sh
./configure --prefix=/tmp/test-issue-305
make install
PATH=/tmp/test-issue-305/bin:$PATH
<now run your test>
```

**Comment by stuhood** [comment]
@wez : Thanks a lot!

I was able to build a binary and manually test it against the relevant query on OSX, and it appears to work as expected: an event doesn't fire for the parent directory, and does fire for the child.

Unfortunately, the test suite we have around watchman is not very extensive yet, so I can't tell you whether there were any negative effects from this change on our end.

**Comment by stuhood** [comment]
@wez : Thanks again; please let me know if there is anything we can do to help get this landed.

**Comment by wez** [comment]
I pushed the code for this, and it will go out in the watchman 4.7 release.  We're not quite ready to cut the release yet.

I'll re-open this issue and update it once we start rolling.

**Comment by wez** [comment]
Tagged 4.7.0 today.  This will be available in homebrew once https://github.com/Homebrew/homebrew-core/pull/4701 is merged.

**Comment by stuhood** [comment]
@wez : Hey Wez! A hopefully quick question to validate my understanding. It appears that in watchman `4.9.0`, neither creates nor updates of files invalidate the parent directory.

A created/deleted child should still invalidate a parent directory, right? This ticket was really just concerned with updates.

**Comment by wez** [comment]
@stuhood creates/updates of files (but looking at the code, apparently not dirs: https://github.com/facebook/watchman/blob/master/root/stat.cpp#L218) should (AFAIR) cause watchman to inspect the parent dir and flag it as changed if the `lstat(2)` data is different since the last time it looked.  Watchman will also update a dir whenever the kernel tells us that the dir has changed, which some systems won't do reliably.

The overwhelming majority of watchman consumers are focused on files more than dirs, so I don't think anyone else has noticed issues in this area or raised them before :-/
